src/Preg.php

```diff
@@ -22,6 +22,35 @@
  */
 final class Preg
 {
+    /**
+     * @param string|string[] $pattern
+     *
+     * @return string|string[]
+     */
+    public static function verify($pattern)
+    {
+        if (\is_array($pattern)) {
+            return \array_map(__METHOD__, $pattern);
+        }
+
+        $last = error_get_last();
+        $result = @preg_match($pattern, '');
+        if (false !== $result) {
+            return $pattern;
+        }
+        $next = error_get_last();
+        if ($last !== $next) {
+            $message = sprintf('(code: %d) %s', preg_last_error(), $next['message']);
+        } else {
+            $message = sprintf('(code: %d)', preg_last_error());
+        }
+
+        throw new PregException(
```

Comment:
i kind of agree with @julienfalque here.
what is a benefit of manual validation rather than letting the `preg_*` validate during normal flow?

in both cases, we throw the `PregException`.

Reply:
Yes, on that level it looks nearly identical. The main reason I tried to highlight in the commit message by writing:

> as getting details of any pattern related errors later on is not easily possible b/c of multi-
matching w/ and w/o UTF-8 mode (see b29848e).

Let me write the little story here:

So when I was executing the CS Fixer, I just was getting PregExceptions which resulted  that the fixer had failures. I could not see which kind of error there was in specific.

And as usual, you get these errors the time you don't need them.

As I'm using the source version I was able to find out w/ step debugging (which was hard as I had the issue w/ PHP 7.4 and my local Xdebug  version with Exception breakpoints in Phpstorm IDE did crash fully, so much for debugging this quick then).

It turned out that it was an invalid pattern. So it would have been good to me for knowing this, but the exact error information was hidden.

Maybe I was too inexperienced and I should have activated some scream first as the the calls suppress warnings (for good reason as two patterns are tested (by the match) for one pattern each).

I first tried to gain the additional information at the place where the exception is thrown (at the end of the function body) as I finally found the spot. And I thought how cruel, throwing an exception with no detailed error message. What would be better?

And then reading the code above I started to realize what made it hard so far to provide a detailed message. It would have certainly required to collect the error information on multiple places. If the pattern is valid first and then invalid this would be good to know, but certainly would introduce many changes, and in the end the issue was with the pattern provided via the parameter.

Now with the suggested changes applied on my box locally I was directly getting the error message and could address the problem.

That detailed error message did allow me in my case to easily spot the bug report in the redhat bugtracker which reminded me I need to update my src version that I had in use as there was a patch already applied and my version was just stale.

In this case I thought, wouldn't it be useful if for project collaborators such errors would be more visible upfront? As the work was already done I just filed this PR.

So this is just the story.

Replacement inputs are not checked (that is similarly possible, too).

So to finally answer your question. This was the benefit to me. A benefit in general is to fail early perhaps, as the pattern is passed stringly typed this is perhaps worth, but I don't need this change upstream. I thought it would be better to have these error messages upfront, but maybe having some contribution / development guidelines that show how to have the application creaming all over the place might have already sufficed as well w/o any further code changes and it's just me who typed too much.

Reply:
I see the value of exposing more details on exception.
I don't see a need to validate regex twice - one on explicit validation, one on internal validation. For me, it's like executing same action twice.
What's more, if John Doe would drop the pre-validation, outcome of the execution would be the same.

Reply:
I think this was an oversight I had in the first push, I changed the visibility to private. Let me double check. Yes, they are private already. Also moved to the end of the class. Refreshing on top of 2.12 now.

Reply:
i'm not talking about visibility of the method - i'm talking about fact that we in fact validate twice now

Reply:
@keradus This did already validate twice on failure before the change-set. So actually this validates only once in case of error (with all due respect: which if I understand you correctly would be in your benefit contrary to what you argue against).

However this could be a resolution. For all of those methods (IIRC four to six), just the error message function could be put to enrich the error message (or take care of the error signalling in full). This would also spare to have the `new` in all those places. It's sort of just a new idea, I can't say if this works out in the end, but it looks not too idiotic.

It's perhaps worth to make it a static method of the exception itself. Please let me know if you have some preferences here or just feedback in general before I change this. Also I would like to know if you're fine to have any of this static function public, not private.

If I touch the PregException class I would turn it final as well if it is not yet (and it does not break too many test-cases).

Reply:
@keradus Validating any of the pattern(s) is now only done in case the main method already error'ed. This should address your comment. I still don't get the build green though, so I might need some more builds until done.